oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/grpc/GRPCCallback.java

```diff
@@ -0,0 +1,144 @@
+
+        GRPCStreamStatus status = new GRPCStreamStatus();
+
+        StreamObserver<org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage> streamObserver =
+            alarmServiceStub.withDeadlineAfter(10, TimeUnit.SECONDS).doAlarm(new StreamObserver<Response>() {
+                @Override
+                public void onNext(Response response) {
+                    // ignore empty response
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+                    status.done();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message failed: {}", throwable.getMessage());
+                    }
+                }
+
+                @Override
+                public void onCompleted() {
+                    status.done();
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send alarm message successful.");
+                    }
+                }
+            });
+
+        alarmMessage.forEach(message -> {
+            org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.Builder builder =
+                org.apache.skywalking.oap.server.core.alarm.grpc.AlarmMessage.newBuilder();
+
+            builder.setScopeId(message.getScopeId());
+            builder.setScope(message.getScope());
+            builder.setName(message.getName());
+            builder.setId0(message.getId0());
+            builder.setId1(message.getId1());
+            builder.setRuleName(message.getRuleName());
+            builder.setAlarmMessage(message.getAlarmMessage());
+            builder.setStartTime(message.getStartTime());
+
+            streamObserver.onNext(builder.build());
+        });
+
+        streamObserver.onCompleted();
+
+        long sleepTime = 0;
+        long cycle = 100L;
+
+        // For memory safe of oap, we must wait for the peer confirmation.
+        while (!status.isDone()) {
+            try {
+                sleepTime += cycle;
+                Thread.sleep(cycle);
+            } catch (InterruptedException ignored) {
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Send {} alarm message to {}:{}.", alarmMessage.size(),
+                          alarmSetting.getTargetHost(), alarmSetting.getTargetPort()
+                );
+            }
+
+            if (sleepTime > 2000L) {
+                log.warn("Send {} alarm message to {}:{}, wait {} milliseconds.", alarmMessage.size(),
+                         alarmSetting.getTargetHost(), alarmSetting.getTargetPort(), sleepTime
+                );
+                cycle = 2000L;
+            }
+        }
+    }
+
+    private void onGRPCAlarmSettingUpdated(GRPCAlarmSetting grpcAlarmSetting) {
+        if (!grpcAlarmSetting.equals(alarmSetting)) {
+            grpcClient.shutdown();
+            grpcClient = new GRPCClient(grpcAlarmSetting.getTargetHost(), grpcAlarmSetting.getTargetPort());
```

Comment:
What if the user delete the gRPC hooks entirely? Say target host and port is empty